File: src/main/java/org/pulcifer/climate/dto/SimilarCity.java
package org.pulcifer.climate.dto;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import org.pulcifer.climate.model.City;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

@JsonInclude(JsonInclude.Include.NON_NULL)
public class SimilarCity {
    private static final MathContext mathContext = new MathContext(4, RoundingMode.HALF_UP);

    @JsonIgnore
    private City city;
    @JsonIgnore
    private Map<ClimateRangeType, List<BigDecimal>> climateData;
    @JsonIgnore
    public List<BigDecimal> getAvgTemperatures() { return climateData.get(ClimateRangeType.AVG_TEMPERATURE); }
    @JsonIgnore
    public List<BigDecimal> getLowestTemperatures() {
        return climateData.get(ClimateRangeType.LOWEST_TEMPERATURE);
    }
    @JsonIgnore
    public List<BigDecimal> getHighestTemperatures() {
        return climateData.get(ClimateRangeType.HIGHEST_TEMPERATURE);
    }
    public enum ScoreTypes {
        RAINFALL, RAIN_DAYS,
        AVG_TEMP, LOWEST_AVG_TEMP, HIGHEST_AVG_TEMP,
        HIGHEST_TEMP, LOWEST_TEMP;
    }

    private Map<ScoreTypes, BigDecimal> ranks;

    @JsonInclude
    public Map<ScoreTypes, BigDecimal> getRanks() {
        return ranks;
    }

    public SimilarCity(City c) {
        this.city = c;
//        this.climateData = new CityClimateBuilder().buildCityClimate(city);
        this.ranks = new HashMap<>();
    }

    public String getCityName() { return city.getCityName(); }
    public Integer getCityId() { return city.getCityId(); }
    public String getCountry() {
        return city.getCountry();
    }
    public Integer getPopulation() {
        return city.getPopulation();
    }
    public String getLatitude() { return city.getCityLatitude(); }
    public String getLongitude() { return city.getCityLongitude(); }
    public BigDecimal getHighestAvgTemperature() {
        if (getAvgTemperatures().isEmpty()) return new BigDecimal("-1");
        return getAvgTemperatures().get(getAvgTemperatures().size() - 1);
    }
    public BigDecimal getLowestAvgTemperature() {
        if (getAvgTemperatures().isEmpty()) return new BigDecimal("-1");
        return getAvgTemperatures().get(0);
    }
    public BigDecimal getAvgTemperature() {
        if (getAvgTemperatures().isEmpty()) return new BigDecimal("-1");
        return getAvgTemperatures().stream()
                .reduce(BigDecimal.ZERO, BigDecimal::add)
                .divide(new BigDecimal(getAvgTemperatures().size()), mathContext);
    }
    public BigDecimal getLowestTemperature() {
        if (getLowestTemperatures().isEmpty()) return new BigDecimal("-1");
        return getLowestTemperatures().get(0);
    }
    public BigDecimal getHighestTemperature() {
        if (getHighestTemperatures().isEmpty()) return new BigDecimal("-1");
        return getHighestTemperatures().get(getHighestTemperatures().size() - 1);
    }
    public BigDecimal getRainDays() { return climateData.get(ClimateRangeType.RAIN_DAYS).stream().reduce(BigDecimal.ZERO, BigDecimal::add); }
    public BigDecimal getRainfall() { return climateData.get(ClimateRangeType.RAINFALL).stream().reduce(BigDecimal.ZERO, BigDecimal::add); }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof SimilarCity)) return false;
        SimilarCity thatSC = (SimilarCity)obj;
        return thatSC.getCityId().equals(getCityId());
    }

    public BigDecimal getTotalRank() {
        BigDecimal score = BigDecimal.ONE;
        for (BigDecimal rank : ranks.values()) {
            if (BigDecimal.ZERO.equals(rank)) continue;
            score = score.add(rank);//.sqrt(mathContext));
        }
        return score;
    }

//    public BigDecimal getTotalRank() {
//        BigDecimal score = BigDecimal.ONE;
//        for (BigDecimal rank : ranks.values()) {
//            if (BigDecimal.ZERO.equals(rank)) continue;
//            score = score.multiply(rank);//.sqrt(mathContext));
//        }
//        return score;
//    }

    public void setRainfallRank(BigDecimal rainfallRank) {
        ranks.put(ScoreTypes.RAINFALL, rainfallRank);
    }

    public void setRainDaysRank(BigDecimal rainDaysRank) {
        ranks.put(ScoreTypes.RAIN_DAYS, rainDaysRank);
    }

    public void setHighestAvgTemperatureRank(BigDecimal highestAvgTemperatureRank) {
        ranks.put(ScoreTypes.HIGHEST_AVG_TEMP, highestAvgTemperatureRank);
    }

    public void setLowestAvgTemperatureRank(BigDecimal lowestAvgTemperatureRank) {
        ranks.put(ScoreTypes.LOWEST_AVG_TEMP, lowestAvgTemperatureRank);
    }

    public void setAvgTemperatureRank(BigDecimal avgTemperatureRank) {
        ranks.put(ScoreTypes.AVG_TEMP, avgTemperatureRank);
    }

    public void setHighestTemperatureRank(BigDecimal highestTemperatureRank) {
        ranks.put(ScoreTypes.HIGHEST_TEMP, highestTemperatureRank);
    }

    public void setLowestTemperatureRank(BigDecimal lowestTemperatureRank) {
        ranks.put(ScoreTypes.LOWEST_TEMP, lowestTemperatureRank);
    }
}

File: src/main/kotlin/org/pulcifer/climate/model/SimilarCity.kt
package org.pulcifer.climate.model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import org.pulcifer.climate.range.ClimateRangeType
import java.math.BigDecimal
import java.util.stream.Collectors
import java.math.MathContext
import java.math.RoundingMode
import java.util.*
import java.util.function.Function

@JsonInclude(JsonInclude.Include.NON_NULL)
class SimilarCity(c: City, diff: Int) {
    @JsonIgnore
    private val city: City

    @JsonIgnore
    val diff: Int

    @JsonIgnore
    private val climateData: Map<ClimateRangeType, List<BigDecimal?>>

    @get:JsonIgnore
    val avgTemperatures: List<BigDecimal?>
        get() = climateData[ClimateRangeType.AVG_TEMPERATURE]!!

    @get:JsonIgnore
    val lowestTemperatures: List<BigDecimal?>
        get() = climateData[ClimateRangeType.LOWEST_TEMPERATURE]!!

    @get:JsonIgnore
    val highestTemperatures: List<BigDecimal?>
        get() = climateData[ClimateRangeType.HIGHEST_TEMPERATURE]!!

    private fun extractListForRange(city: City, getValue: Function<ClimateMonth?, BigDecimal?>): List<BigDecimal?> {
        val climateMonths = city.climate?.climateMonth
        val values = climateMonths!!.stream().map { m: ClimateMonth? -> getValue.apply(m) }.collect(Collectors.toList())
        return values.sortedBy{it}
    }

    val cityId: Int?
        get() = city.cityId
    val cityName: String?
        get() = city.cityName
    val country: String?
        get() = city.country
    val highestAvgTemperature: BigDecimal?
        get() = getMax(avgTemperatures)
    val lowestAvgTemperature: BigDecimal?
        get() = getMin(avgTemperatures)
    val avgTemperature: BigDecimal
        get() = if (avgTemperatures.isEmpty()) BigDecimal.ZERO else avgTemperatures.stream()
            .reduce(BigDecimal.ZERO) { obj: BigDecimal?, augend: BigDecimal? -> obj!!.add(augend) }!!
            .divide(
                BigDecimal(
                    avgTemperatures.size
                ), context
            )
    val lowestTemperature: BigDecimal?
        get() = getMin(lowestTemperatures)
    val highestTemperature: BigDecimal?
        get() = getMax(highestTemperatures)
    val rainDays: BigDecimal?
        get() = climateData[ClimateRangeType.RAIN_DAYS]!!
            .stream().reduce(BigDecimal.ZERO) { obj: BigDecimal?, augend: BigDecimal? -> obj!!.add(augend) }
    val rainfall: BigDecimal?
        get() = climateData[ClimateRangeType.RAINFALL]!!
            .stream().reduce(BigDecimal.ZERO) { obj: BigDecimal?, augend: BigDecimal? -> obj!!.add(augend) }

    private fun getMax(l: List<BigDecimal?>): BigDecimal? {
        if (l.isEmpty()) return BigDecimal.ONE
        Collections.reverse(l)
        return l[0]
    }

    private fun getMin(l: List<BigDecimal?>): BigDecimal? {
        if (l.isEmpty()) return BigDecimal.ONE
        return l[0]
    }

    companion object {
        private val context = MathContext(4, RoundingMode.HALF_UP)
        private val TWO = BigDecimal("2")
    }

    init {
        val climateData: MutableMap<ClimateRangeType, List<BigDecimal?>> = HashMap()
        climateData[ClimateRangeType.AVG_TEMPERATURE] = extractListForRange(c) { m: ClimateMonth? ->
            m?.minTempF?.add(m.maxTempF)?.divide(
                TWO, context
            )
        }
        climateData[ClimateRangeType.LOWEST_TEMPERATURE] =
            extractListForRange(c) { m: ClimateMonth? -> m?.minTempF }
        climateData[ClimateRangeType.HIGHEST_TEMPERATURE] =
            extractListForRange(c) { m: ClimateMonth? -> m?.maxTempF }
        climateData[ClimateRangeType.RAIN_DAYS] = extractListForRange(c) { m: ClimateMonth? -> m?.raindays }
        climateData[ClimateRangeType.RAINFALL] = extractListForRange(c) { m: ClimateMonth? -> m?.rainfall }
        this.city = c
        this.diff = diff
        this.climateData = climateData
    }
}